### PR TITLE
Sortable list animations

### PR DESCRIPTION
**Pull request by AtomicAlienZ**
### 🔦 Summary
TBD

### ✅ Checklist
- [ ] 👨‍💻 API change is approved by the UI Infra Developer 
- [ ] 👨‍🎨 UX change is approved by the Design System UX 
- [ ] 📚 Change is documented
  - [ ] Story
  - [ ] API description
  - [ ] Other (explain)
- [ ] 🔬 Change is tested
  - [ ] Component
  - [ ] Visual test

### CODE REVIEW

**Comment by AtomicAlienZ** [comment]
Perfomance is not so great but manageable, could be improved by shifting nodes via direct node access, not passing props to subcomponents

File: src/DragAndDrop/Draggable/Draggable.js [hunk 1]
```diff
@@ -40,6 +41,14 @@ Draggable.propTypes = {
   onDragStart: PropTypes.func,
   /** callback for drag end */
   onDragEnd: PropTypes.func,
+
+  shift: PropTypes.arrayOf(PropTypes.number),
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/Draggable.js]: what is shift?
i think it make sense to left comments about this

File: src/DragAndDrop/Draggable/Draggable.js [hunk 2]
```diff
@@ -40,6 +41,14 @@ Draggable.propTypes = {
...
+  hasDragged: PropTypes.bool,
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/Draggable.js]: what mean hasDragged? that drag just done?

File: src/DragAndDrop/Draggable/Draggable.js [hunk 3]
```diff
@@ -40,6 +41,14 @@ Draggable.propTypes = {
...
+  setWrapperNode: PropTypes.func,
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/Draggable.js]: same here, not clear for me what is doing)

File: src/DragAndDrop/Draggable/Draggable.js [hunk 4]
```diff
@@ -40,6 +41,14 @@ Draggable.propTypes = {
...
+  /** animation duration in ms, default = 0 - disabled */
+  animationDuration: PropTypes.number,
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/Draggable.js]: should we have separete prop to enable/disable animation? wdyt?
@shlomitc

File: src/DragAndDrop/Draggable/components/DraggableSource.js [hunk 5]
```diff
@@ -153,16 +180,20 @@ export default class DraggableSource extends React.Component {
...
         id,
         item,
         isPlaceholder: isDragging,
         connectHandle: noop,
-      }),
+      }))
+    );
+
+    return (
+      <div style={this._getWrapperStyles()}>
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/components/DraggableSource.js]: @shlomitc should we expose this styles to user, in renderItem or we can just have extra node here?

File: src/DragAndDrop/Draggable/components/DraggableSource.js [hunk 6]
```diff
@@ -218,4 +249,11 @@ DraggableSource.propTypes = {
   onMoveOut: PropTypes.func,
   onDragStart: PropTypes.func,
   onDragEnd: PropTypes.func,
+
+  shift: PropTypes.arrayOf(PropTypes.number),
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/components/DraggableSource.js]: .

File: src/SortableList/SortableList.js [hunk 7]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
-      this.setState({ items });
-    }
+  wrapperNodes = [];
+
+  setWrapperNode = (node, index, item) => {
+    this.wrapperNodes[index] = {node, index, item};
+  };
+
+  componentWillReceiveProps({items}) {
+    this.setState(prevState => ({
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: do we need to do setState on each call?

File: src/SortableList/SortableList.js [hunk 8]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
+      placeholderShift: [0, 0],
+      isDragging: null,
+    }));
   }
 
   handleMoveOut = id => {
-    this.setState({ items: this.state.items.filter(it => it.id !== id) });
+    this.setState({
+      items: this.state.items.filter(it => it.id !== id),
+      animationShifts: {},
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: maybe make sense to move this animation stuff to separate method, like resetAnimationState or smth like this

File: src/SortableList/SortableList.js [hunk 9]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
+    });
+    this.wrapperNodes = this.wrapperNodes.filter(({item}) => item.id !== id);
   };
 
-  handleHover = (removedIndex, addedIndex, options = {}) => {
-    this.setState(prevState => {
-      const nextItems = [...prevState.items];
-      if (!nextItems.find(it => it.id === options.id)) {
-        nextItems.splice(addedIndex, 0, options.item);
+  getAnimationShifts = (originalIndex, moveToIndex, shiftIndex) => {
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: this method defenetly need to be documented, and maybe even move out from a component

File: src/SortableList/SortableList.js [hunk 10]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
+      }
+    }
+
+    return [shiftX, shiftY];
+  };
+
+  handleHover = ({addedIndex, item}) => {
+    this.setState(prevState => {
+      const originalIndex = this.state.items.indexOf(item);
+      const items = [...prevState.items];
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: prevState.items.slice() - faster)

File: src/SortableList/SortableList.js [hunk 11]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
+      let animationShifts = {};
+      let isDragging = this.state.isDragging;
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: { isDragging }

File: src/SortableList/SortableList.js [hunk 12]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
+
+      // New item added from other list
+      if (originalIndex < 0) {
+        items.splice(addedIndex, 0, item);
+        isDragging = item.id;
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: boolean !== string

File: src/SortableList/SortableList.js [hunk 13]
```diff
@@ -2,38 +2,139 @@ import React from 'react';
...
+
+      if (placeholderRect.x === targetRect.x) {
+        shiftY = shiftForward ? (targetRect.bottom - placeholderRect.bottom) : (targetRect.top - placeholderRect.top);
+      }
+    }
+
+    return [shiftX, shiftY];
+  };
+
+  handleHover = ({addedIndex, item}) => {
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: require docs

File: src/SortableList/SortableList.js [hunk 14]
```diff
@@ -54,19 +155,21 @@ export default class SortableList extends WixComponent {
   };
 
   handleDragStart = data => {
+    this.setState({animationShifts: {}, isDragging: data.id, placeholderShift: [0, 0]});
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: we did this in componentWillReceiveProps, do not we?

File: src/SortableList/SortableList.js [hunk 15]
```diff
@@ -108,14 +222,21 @@ export default class SortableList extends WixComponent {
           <div className={contentClassName}>
             {this.state.items.map((item, index) => (
               <Draggable
+                key={`${item.id}-${containerId}`}
+
+                shift={this.state.animationShifts[index]}
+                hasDragged={!!this.state.isDragging && this.state.isDragging !== item.id}
```
Comment (Faradey27) [inline src/SortableList/SortableList.js]: ```this.state.isDragging !== item.id``` - we can better i think)

File: stories/DragAndDrop/Introduction/Introduction.md [hunk 16]
```diff
@@ -12,3 +12,11 @@ The most useful building block you should be aware of is the `<SortableList/>` c
 
 ### Example
 Here is a simple usage of the `<SortableList/>` component with basic styles (Give it a try)!
+
+### Animations
+
+#### Default animation effect
+By default there is an animation for SortableList items positions. Component does not updates its state related to items positions while dragging, but shows you a future state of items animated to new position.
```
Comment (Faradey27) [inline stories/DragAndDrop/Introduction/Introduction.md]: i do not undestand))

File: stories/DragAndDrop/Introduction/IntroductionExample.js [hunk 17]
```diff
@@ -17,15 +17,31 @@ export default class IntroductionExample extends React.Component {
         },
         {
           id: 'b',
-          text: 'Item 2',
+          text: 'Item 2Item 2',
```
Comment (Faradey27) [inline stories/DragAndDrop/Introduction/IntroductionExample.js]: maybe better to create separate story for animations?
@shlomitc

File: stories/DragAndDrop/Introduction/IntroductionExample.scss [hunk 18]
```diff
@@ -2,7 +2,7 @@
   box-sizing: border-box;
   border-radius: 8px;
   width: 400px;
-  height: 45px;
+  //height: 45px;
```
Comment (Faradey27) [inline stories/DragAndDrop/Introduction/IntroductionExample.scss]: .

File: stories/DragAndDrop/SortableList/MultiAreaListWithSortableColumns/MultiAreaListWithSortableColumns.js [hunk 19]
```diff
@@ -27,10 +27,17 @@ export default class MultiAreaListWithSortableColumns extends React.Component {
       {
         id: 'multiArea1',
         items: generateStateForContainer(4, 1),
+        w: 250,
```
Comment (Faradey27) [inline stories/DragAndDrop/SortableList/MultiAreaListWithSortableColumns/MultiAreaListWithSortableColumns.js]: what does it mean?

File: stories/DragAndDrop/SortableList/MultiAreaListWithSortableColumns/MultiAreaListWithSortableColumns.js [hunk 20]
```diff
@@ -99,7 +106,7 @@ export default class MultiAreaListWithSortableColumns extends React.Component {
     );
 
     return (
-      <div className={classes} style={previewStyles} data-hook={`column-${id}`}>
+      <div className={classes} style={{...previewStyles, width: item.w}} data-hook={`column-${id}`}>
```
Comment (Faradey27) [inline stories/DragAndDrop/SortableList/MultiAreaListWithSortableColumns/MultiAreaListWithSortableColumns.js]: oh, width)

File: stories/DragAndDrop/Introduction/IntroductionExample.js (hunk 17 above)
Comment (AtomicAlienZ) [reply stories/DragAndDrop/Introduction/IntroductionExample.js]: Or we can just add another example to existing stories

File: stories/DragAndDrop/Introduction/Introduction.md (hunk 16 above)
Comment (AtomicAlienZ) [reply stories/DragAndDrop/Introduction/Introduction.md]: Yeah, that needs rephrasing.

File: src/SortableList/SortableList.js (hunk 7 above)
Comment (AtomicAlienZ) [reply src/SortableList/SortableList.js]: Yes. This is the only way we've found to clear state on drag end if element was dragged from another list. Will add a comment that explains this.

File: src/SortableList/SortableList.js (hunk 12 above)
Comment (AtomicAlienZ) [reply src/SortableList/SortableList.js]: Naming issue. Good catch.

File: src/SortableList/SortableList.js (hunk 9 above)
Comment (cTux) [reply src/SortableList/SortableList.js]: Documented, not sure if it should be elsewhere except this component.

File: src/DragAndDrop/Draggable/Draggable.js (hunk 1 above)
Comment (cTux) [reply src/DragAndDrop/Draggable/Draggable.js]: Shift (or offset) - visual positioning shifting for an element (transform: translate) without moving it from its real position at DOM.

Comment (cTux) [reply src/DragAndDrop/Draggable/Draggable.js]: Description is added to components.

File: src/DragAndDrop/Draggable/Draggable.js (hunk 4 above)
Comment (cTux) [reply src/DragAndDrop/Draggable/Draggable.js]: `animationDuration` by default is 0, which means no animation at all

File: src/DragAndDrop/Draggable/components/DraggableSource.js (hunk 6 above)
Comment (cTux) [reply src/DragAndDrop/Draggable/components/DraggableSource.js]: Description is added.

File: src/SortableList/SortableList.js (hunk 9 above)
Comment (cTux) [reply src/SortableList/SortableList.js]: Or do you mean some util module?

File: stories/DragAndDrop/Introduction/IntroductionExample.js (hunk 17 above)
Comment (shlomitc) [reply stories/DragAndDrop/Introduction/IntroductionExample.js]: add a new story anyway

File: stories/DragAndDrop/Introduction/IntroductionAnimatedExample.js [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+
+  handleDrop = ({ removedIndex, addedIndex }) => {
+    const nextItems = [...this.state.items];
+    nextItems.splice(addedIndex, 0, ...nextItems.splice(removedIndex, 1));
+    this.setState({
+      items: nextItems,
+    });
+  };
+
+  renderItem = ({ isPlaceholder, isPreview, id, previewStyles, item }) => {
```
Comment (xmityaz) [inline stories/DragAndDrop/Introduction/IntroductionAnimatedExample.js]: @AtomicAlienZ @cTux 
Is `previewStyles` relevant here?
Looks like it doesn't come from `renderItem` method arguments

File: src/DragAndDrop/Draggable/components/DraggableSource.js [hunk 22]
```diff
@@ -78,6 +78,17 @@ const source = {
...
+    { id, index, containerId, groupName, item, canDrag },
+  ) => {
+    return canDrag ? canDrag({
+      id,
+      index,
+      containerId,
+      groupName,
+      item,
+    }) : true;
+  },
```
Comment (Dariia) [inline src/DragAndDrop/Draggable/components/DraggableSource.js]: @Faradey27 Looks like shouldCancelSortStart  from issue https://github.com/wix/wix-style-react/issues/2793  implemented as canDrag here

**Pull request merged**